# Dimensiones de calidad de datos

## 1. Exploración inicial

In [433]:
""" 1.1 Importacion de librerias de ciencia de datos """
import pandas as pd
import numpy as np
import re

secop_df = pd.read_parquet('secop_bienes.parquet')

In [434]:
""" 1.2 Dimensiones del dataset, Revisión de la cantidad de filas y columnas para dimensionar el volumen de datos."""
print('Filas:', secop_df.shape[0])
print('Columnas:', secop_df.shape[1])

Filas: 196391
Columnas: 36


In [435]:
""" 1.3 Obtencion de los nombres de las columnas. """
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)
secop_df.columns

Index(['id_contrato', 'nombre_entidad', 'departamento', 'ciudad', 'orden',
       'sector', 'rama', 'entidad_centralizada', 'tipo_de_contrato',
       'modalidad_de_contratacion', 'codigo_de_categoria_principal',
       'objeto_del_contrato', 'fecha_de_firma', 'fecha_de_inicio_del_contrato',
       'fecha_de_fin_del_contrato', 'duraci_n_del_contrato',
       'condiciones_de_entrega', 'proveedor_adjudicado', 'documento_proveedor',
       'es_pyme', 'es_grupo', 'g_nero_representante_legal',
       'nacionalidad_representante_legal', 'valor_del_contrato',
       'valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago',
       'dias_adicionados', 'habilita_pago_adelantado',
       'el_contrato_puede_ser_prorrogado', 'origen_de_los_recursos',
       'destino_gasto', 'estado_contrato', 'liquidaci_n',
       'obligaci_n_ambiental', 'urlproceso'],
      dtype='str')

In [436]:
""" 1.4 Mostrar las primeras 5 filas del DataFrame. """
secop_df.head()

,id_contrato,nombre_entidad,departamento,ciudad,orden,sector,rama,entidad_centralizada,tipo_de_contrato,modalidad_de_contratacion,codigo_de_categoria_principal,objeto_del_contrato,fecha_de_firma,fecha_de_inicio_del_contrato,fecha_de_fin_del_contrato,duraci_n_del_contrato,condiciones_de_entrega,proveedor_adjudicado,documento_proveedor,es_pyme,es_grupo,g_nero_representante_legal,nacionalidad_representante_legal,valor_del_contrato,valor_pagado,valor_facturado,valor_pendiente_de_pago,dias_adicionados,habilita_pago_adelantado,el_contrato_puede_ser_prorrogado,origen_de_los_recursos,destino_gasto,estado_contrato,liquidaci_n,obligaci_n_ambiental,urlproceso
0,CO1.PCCNTR.1000001,PARQUES NACIONALES NATURALES DE COLOMBIA - DIRECCION TERRITORIAL AMAZONIA,Distrito Capital de Bogotá,No Definido,Territorial,Ambiente y Desarrollo Sostenible,Ejecutivo,Centralizada,Suministros,Mínima cuantía,V1.50161509,Contratar a monto agotable por el sistema de precios unitarios; sin formula de reajuste el suministro de elementos de aseo y limpieza e insumos de cafetería y restaurante; para los Parques Nacionales Naturales de Amacayacu; Cahuinarí; Río Puré y Yaigojé Apaporis,2019-06-19,2019-06-18,2019-12-31,No definido,Como acordado previamente,INDUSTRIAS GUERRERO Y COMPAÑIA S.A.S.,830130048,Si,No,No Definido,CO,17897916.0,17896237,17896237,1679,0,No Definido,Si,Distribuido,Funcionamiento,Cerrado,Si,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.846428&isFromPublicArea=True&isModal=true&asPopupView=true'}
1,CO1.PCCNTR.1000507,AGENCIA LOGISTICA DE LAS FUERZAS MILITARES,Distrito Capital de Bogotá,Bogotá,Nacional,defensa,Ejecutivo,Centralizada,Suministros,Selección Abreviada de Menor Cuantía,V1.50131612,SUMINISTRO DE HUEVOS CON DESTINO A LOS COMEDORES DE TROPA DE LA BR-4 Y BR-14; ADMINISTRADOS POR LA AGENCIA LOGÍSTICA DE LAS FUERZAS MILITARES REGIONAL ANTIOQUIA CHOCO Y OTRAS POSIBLES UNIDADES QUE LO REQUIERAN,2019-06-23,2019-06-13,2019-11-29,Dia(s),Transporte incluido,Distribuidora Antioqueña AM,21788564,Si,No,Mujer,CO,390000000.0,389999685,389999685,315,0,No,No,Distribuido,Funcionamiento,Cerrado,Si,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.841446&isFromPublicArea=True&isModal=true&asPopupView=true'}
2,CO1.PCCNTR.1000901,FUERZA AEROESPACIAL COLOMBIANA,Distrito Capital de Bogotá,Bogotá,Nacional,defensa,Ejecutivo,Centralizada,Compraventa,Mínima cuantía,V1.53102701,ADQUISICIÓN DE PRESILLAS BORDADAS AZULES EN CANUTILLO PARA OFICIALES DE INSIGNIA DE LA FUERZA AÉREA COLOMBIANA CONFORME ANEXO TECNICO,2019-06-18,2019-06-24,2019-09-30,No definido,DAP - Entregado en un punto (lugar de destino convenido),FANNY JANETH GUTIERREZ PRIETO,20484821,No,No,Mujer,CO,11970000.0,11970000,11970000,0,0,No Definido,Si,Distribuido,Funcionamiento,terminado,No,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.846657&isFromPublicArea=True&isModal=true&asPopupView=true'}
3,CO1.PCCNTR.1000908,AGENCIA LOGISTICA DE LAS FUERZAS MILITARES,Distrito Capital de Bogotá,Bogotá,Nacional,defensa,Ejecutivo,Centralizada,Compraventa,Licitación pública,V1.41115300,ADQUISICIÓN DE UN SISTEMA DE SONAR 3D DE ALTA RESOLUCIÓN CON DESTINO AL DEPARTAMENTO DE BUCEO Y SALVAMENTO,2019-07-25,2019-07-12,2019-11-30,Dia(s),Como acordado previamente,CASCO ANTIGUO,76044753,No,No,Otro,CO,733992000.0,0,0,733992000,0,No,No,Distribuido,Funcionamiento,Cerrado,Si,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.826239&isFromPublicArea=True&isModal=true&asPopupView=true'}
4,CO1.PCCNTR.1000911,PARQUES NACIONALES NATURALES DE COLOMBIA - DIRECCION TERRITORIAL CARIBE,Bolívar,Cartagena,Territorial,Ambiente y Desarrollo Sostenible,Ejecutivo,Centralizada,Suministros,Mínima cuantía,V1.92121701,Suministro de equipos para la instalación de todas las cámaras necesarias para el monitoreo; para ser instalados en\nlas sedes oper

In [437]:
""" 1.5 Mostrar las ultimas 5 filas del DataFrame. """
secop_df.tail

<bound method NDFrame.tail of                id_contrato  \
0       CO1.PCCNTR.1000001   
1       CO1.PCCNTR.1000507   
2       CO1.PCCNTR.1000901   
3       CO1.PCCNTR.1000908   
4       CO1.PCCNTR.1000911   
...                    ...   
196386   CO1.PCCNTR.999805   
196387   CO1.PCCNTR.999806   
196388   CO1.PCCNTR.999807   
196389   CO1.PCCNTR.999810   
196390   CO1.PCCNTR.999902   

                                                                    nombre_entidad  \
0       PARQUES NACIONALES NATURALES DE COLOMBIA - DIRECCION  TERRITORIAL AMAZONIA   
1                                       AGENCIA LOGISTICA DE LAS FUERZAS MILITARES   
2                                                   FUERZA AEROESPACIAL COLOMBIANA   
3                                       AGENCIA LOGISTICA DE LAS FUERZAS MILITARES   
4          PARQUES NACIONALES NATURALES DE COLOMBIA - DIRECCION TERRITORIAL CARIBE   
...                                                                            ...   
196386 

## 2. Validez

In [438]:
""" 2.1 Análisis de validez de secop_df """

print("Tipos de datos actuales:")
display(secop_df.dtypes)



Tipos de datos actuales:


id_contrato                                    str
nombre_entidad                                 str
departamento                                   str
ciudad                                         str
orden                                          str
sector                                         str
rama                                           str
entidad_centralizada                           str
tipo_de_contrato                               str
modalidad_de_contratacion                      str
codigo_de_categoria_principal                  str
objeto_del_contrato                            str
fecha_de_firma                      datetime64[us]
fecha_de_inicio_del_contrato        datetime64[us]
fecha_de_fin_del_contrato           datetime64[us]
duraci_n_del_contrato                          str
condiciones_de_entrega                         str
proveedor_adjudicado                           str
documento_proveedor                            str
es_pyme                        

In [439]:
""" 2.2 Conteo de tipo de fechas """

print(secop_df["fecha_de_firma"].map(type).value_counts(), "\n")
print(secop_df["fecha_de_inicio_del_contrato"].map(type).value_counts(), "\n")
print(secop_df["fecha_de_fin_del_contrato"].map(type).value_counts(), "\n")


fecha_de_firma
<class 'pandas.Timestamp'>    196391
Name: count, dtype: int64 

fecha_de_inicio_del_contrato
<class 'pandas.Timestamp'>             193951
<class 'pandas.api.typing.NaTType'>      2440
Name: count, dtype: int64 

fecha_de_fin_del_contrato
<class 'pandas.Timestamp'>    196391
Name: count, dtype: int64 



In [440]:
""" 2.3 Convertir temporalmente a string solo para evaluar el formato de las fechas """
fechas_str = secop_df["fecha_de_firma"].astype(str)

condiciones = [
    fechas_str.str.contains(r"^[0-9]{4}-[0-9]{2}-[0-9]{2}$", regex=True, na=False),
    fechas_str.str.contains(r"^[0-9]{2}/[0-9]{2}/[0-9]{4}$", regex=True, na=False),
    fechas_str.str.contains(r"^[0-9]{2}-[0-9]{2}-[0-9]{4}$", regex=True, na=False),
    fechas_str.str.contains(r"[a-zA-Z]", regex=True, na=False),
]

etiquetas = [
    "Formato: YYYY-MM-DD",
    "Formato: DD/MM/YYYY",
    "Formato: DD-MM-YYYY",
    "Formato: con mes en texto",
]

patron_detectado = np.select(
    condiciones, etiquetas, default="Formato: otro/desconocido"
)

In [441]:
""" 2.4 Agrupar todos los formatos de fecha_de_inicio_del_contrato que hay en el dataset y su conteo. """


# 4. Agrupamos y construimos el resumen deseado
resultado = (
    secop_df.assign(patron_detectado=patron_detectado)
    .groupby("patron_detectado")
    .agg(
        cantidad=("fecha_de_inicio_del_contrato", "count"),
        ejemplo1=("fecha_de_inicio_del_contrato", "min"),
        ejemplo2=("fecha_de_inicio_del_contrato", "max"),
    )
    .reset_index()
    .sort_values(by="cantidad", ascending=False)
)

# Mostrar la tabla final
resultado



,patron_detectado,cantidad,ejemplo1,ejemplo2
0,Formato: YYYY-MM-DD,193951,2016-06-27,2026-12-16


In [442]:
""" 2.5 Agrupar todos los formatos de fecha_de_fin_del_contrato que hay en el dataset y su conteo. """

resultado = (
    secop_df.assign(patron_detectado=patron_detectado)
    .groupby("patron_detectado")
    .agg(
        cantidad=("fecha_de_fin_del_contrato", "count"),
        ejemplo1=("fecha_de_fin_del_contrato", "min"),
        ejemplo2=("fecha_de_fin_del_contrato", "max"),
    )
    .reset_index()
    .sort_values(by="cantidad", ascending=False)
)

# Mostrar la tabla final
resultado



,patron_detectado,cantidad,ejemplo1,ejemplo2
0,Formato: YYYY-MM-DD,196391,202-12-31,2065-01-31


In [443]:
""" 2.6 Agrupar todos los formatos de fecha_de_firma que hay en el dataset y su conteo. """

resultado = (
    secop_df.assign(patron_detectado=patron_detectado)
    .groupby("patron_detectado")
    .agg(
        cantidad=("fecha_de_firma", "count"),
        ejemplo1=("fecha_de_firma", "min"),
        ejemplo2=("fecha_de_firma", "max"),
    )
    .reset_index()
    .sort_values(by="cantidad", ascending=False)
)

# Mostrar la tabla final
resultado



,patron_detectado,cantidad,ejemplo1,ejemplo2
0,Formato: YYYY-MM-DD,196391,2019-01-01,2025-12-31


### Hallazgos

- El 100% de los registros de fecha_de_inicio_del_contrato, fecha_de_fin_del_contrato y fecha_de_firma tienen el mismo formato de fecha YYYY-MM-DD

## 3. Completitud

In [444]:
""" 3.1 Analisis de completitud de los datos de todas las columnas del DataFrame."""

total_instancias = len(secop_df)
print(f"Total de registros: {total_instancias:,}")


def analizar_completitud(df):
    # Nulos y strings vacíos/blancos, son tratados de igual manera como valores incompletos.
    filas = []
    for columna_test in secop_df.columns:
        serie = secop_df[columna_test]
        es_dato_nulo = serie.isnull()
        if serie.dtype == object:
            es_dato_nulo = es_dato_nulo | (serie.astype(str).str.strip() == "")
        nulos = int(es_dato_nulo.sum())
        no_nulos = total_instancias - nulos
        

        filas.append({
            "columna": columna_test,
            "Total Registros": total_instancias,
            "Valores Completos": no_nulos,
            "Valores Nulos o Vacios": nulos,
         
        })

    df_completitud = pd.DataFrame(filas)
    return df_completitud

df_completitud = analizar_completitud(secop_df)


print("\nRESUMEN DE COMPLETITUD")
display(df_completitud)



df_incompletas = df_completitud[df_completitud["Valores Nulos o Vacios"] >= 1].copy()
df_incompletas["% Nulos o Vacios"] = (
    df_incompletas["Valores Nulos o Vacios"] / total_instancias
) * 100


df_incompletas["% Nulos o Vacios"] = df_incompletas["% Nulos o Vacios"].round(4)

print("\nRESUMEN DE COLUMNAS CON NULOS O VACÍOS (>= 1)")
display(df_incompletas)






Total de registros: 196,391

RESUMEN DE COMPLETITUD


,columna,Total Registros,Valores Completos,Valores Nulos o Vacios
0,id_contrato,196391,196391,0
1,nombre_entidad,196391,196391,0
2,departamento,196391,196391,0
3,ciudad,196391,196391,0
4,orden,196391,196391,0
5,sector,196391,196391,0
6,rama,196391,196391,0
7,entidad_centralizada,196391,196391,0
8,tipo_de_contrato,196391,196391,0
9,modalidad_de_contratacion,196391,196391,0



RESUMEN DE COLUMNAS CON NULOS O VACÍOS (>= 1)


,columna,Total Registros,Valores Completos,Valores Nulos o Vacios,% Nulos o Vacios
13,fecha_de_inicio_del_contrato,196391,193951,2440,1.2424
18,documento_proveedor,196391,196387,4,0.0020


In [445]:
"""3.3 Filtrar registros con documento_proveedor nulo o vacío. """
filtro = secop_df["documento_proveedor"].isna() | (secop_df["documento_proveedor"].astype(str).str.strip() == "")

# Mostrar la tabla completa con esos registros
registros_nulos = secop_df[filtro]

print(f"Se encontraron {len(registros_nulos):,} registros con documento_proveedor nulo o vacío")
display(registros_nulos)


Se encontraron 4 registros con documento_proveedor nulo o vacío


,id_contrato,nombre_entidad,departamento,ciudad,orden,sector,rama,entidad_centralizada,tipo_de_contrato,modalidad_de_contratacion,codigo_de_categoria_principal,objeto_del_contrato,fecha_de_firma,fecha_de_inicio_del_contrato,fecha_de_fin_del_contrato,duraci_n_del_contrato,condiciones_de_entrega,proveedor_adjudicado,documento_proveedor,es_pyme,es_grupo,g_nero_representante_legal,nacionalidad_representante_legal,valor_del_contrato,valor_pagado,valor_facturado,valor_pendiente_de_pago,dias_adicionados,habilita_pago_adelantado,el_contrato_puede_ser_prorrogado,origen_de_los_recursos,destino_gasto,estado_contrato,liquidaci_n,obligaci_n_ambiental,urlproceso
19847,CO1.PCCNTR.1802368,DEPARTAMENTO DE ANTIOQUIA//,Antioquia,Medellín,Territorial,Servicio Público,Ejecutivo,Centralizada,Suministros,Selección abreviada subasta inversa,V1.12352104,COMPRA DE RON 3 AÑOS DE AÑEJAMIENTO DE GRADO MINIMODE 75% V/V,2020-09-11,2020-09-11,2020-12-31,90 Dia(s),Como acordado previamente,UNION TEMPORAL RON PANAMA 2020,NaN,No,Si,No Definido,CO,6.374149e+10,0,0,63741494232,0,No Definido,No,Distribuido,Funcionamiento,terminado,Si,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.1251670&isFromPublicArea=True&isModal=true&asPopupView=true'}
71874,CO1.PCCNTR.4117029,GOBERNACION DEL DEPARTAMENTO DEL CESAR,Cesar,Valledupar,Territorial,Servicio Público,Ejecutivo,Descentralizada,Suministros,Selección abreviada subasta inversa,V1.44103000,SUMINISTRO; INSTALACION; RECARGAS; REPUESTOS; REPARACION; MANTENIMIENTO PREVENTIVO Y CORRECTIVO DE LOS TONERES Y TINTAS ORIGINALES; EQUIPOS DE COMPUTO; EQUIPOS ACTIVOS DE RED; EQUIPOS DE OFICINA IMPRESORAS; FOTOCOPIADORAS; ESCANER; DATACENTER; EQUIPOS TECNOLOGICOS; UPS; SERVIDORES Y CABLEADO ESTRUCTURADO; PERTENECIENTES A LA GOBERNACION DEL DEPARTAMENTO DEL CESAR Y SUS DEPENDENCIAS,2022-10-14,2022-10-20,2022-12-30,78 Dia(s),No Definido,CONSORCIO TECNOSUMINISTROS & SERVICIOS DEL CESAR 2022,NaN,Si,Si,Hombre,CO,3.091562e+09,0,0,3091561509,0,No Definido,No,Distribuido,Inversión,En ejecución,Si,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.3315881&isFromPublicArea=True&isModal=true&asPopupView=true'}
109414,CO1.PCCNTR.5593886,CENTRAL ADMINISTRATIVA ESPECIALIZADA INGENIEROS CENAC INGENIEROS,Distrito Capital de Bogotá,Bogotá,Nacional,defensa,Ejecutivo,Centralizada,Compraventa,Licitación pública,V1.23153206,ADQUISICIÓN DEL ROBOT ANTIEXPLOSIVOS PLATAFORMA ROBÓTICA ANTIEXPLOSIVOS VEHICULOS TERRESTRES NO TRIPULADOS PARA RESPUESTA EOD No JEMPP CEDE10 DINCO COING CENAM ET 03085 ING 4,2023-12-14,2023-12-15,2024-04-20,25 Dia(s),NXTWY.DLVY.2,UNION TEMPORAL ROBOTEAM,NaN,No,Si,Hombre,CO,3.450500e+09,0,0,3450500000,0,No,No,Distribuido,Funcionamiento,terminado,No,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.5077917&isFromPublicArea=True&isModal=true&asPopupView=true'}
133958,CO1.PCCNTR.6704363,Rama Judicial – Dirección Seccional de Administración Judicial de Bucaramanga,Santander,Bucaramanga,Nacional,Ley de Justicia,Judicial,Centralizada,Compraventa,Selección abreviada subasta inversa,V1.44103103,Contratar en nombre de la Nación - Consejo Superior de Judicatura - Dirección Ejecutiva Seccional de Administración Judicial de Bucaramanga; la adquisición de consumibles de impresión para los Despachos Judiciales y Oficinas administrativas de la Seccional Santander; en las condiciones técnicas; de calidad y cantidades requeridas por la entidad,2024-09-16,2024-09-20,2024-10-15,30 Dia(s),NXTWY.DLVY.3,UT TONERS MAXPRINTOTALY-2024,NaN,Si,Si,Hombre,CO,4.145611e+08,0,0,414561126,0,No,Si,Distribuido,Funcionamiento,Cerrado,Si,Si,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.6548090&isFromPublicArea=True&isModal=true&asPopupView=true'}


In [446]:
"""3.2 Top 5 proveedores con más contratos. """
total_registros = len(secop_df)

# Agrupar por proveedor y contar contratos
top_proveedores = (
    secop_df
    .groupby(["documento_proveedor", "proveedor_adjudicado"])
    .size()
    .reset_index(name="cantidad_contratos")
    .sort_values("cantidad_contratos", ascending=False)
)


# Mostrar Top 5
top_5 = top_proveedores.head(5)

print(f"Total de registros: {total_registros:,}")
print("\n📊 Top 5 proveedores con más contratos:")
display(top_5)



Total de registros: 196,391

📊 Top 5 proveedores con más contratos:


,documento_proveedor,proveedor_adjudicado,cantidad_contratos
9231,804003744,CENTRAL DE SUMINISTROS SAS,640
6605,59706955,Deicy Bravo Jojoa,579
11718,860002400,LA PREVISORA S.A. COMPAÑÍA DE SEGUROS,560
10345,828002423,DISTRIBUIDORA COLOMBIANA DE MEDICAMENTOS Y TECNOLOGIAS EN SALUD S.A.S. BIC - DISCOLMETS S.A.S. BIC,475
16756,900503041,SUPRISA S.A.S,471


In [447]:
""" 3.4 Consolidado anual de valores pagados, facturados y pendientes de pago. """

# 1. Asegurar que la columna de fecha esté en formato datetime
secop_df['fecha_de_firma'] = pd.to_datetime(secop_df['fecha_de_firma'])

# 2. Agrupar por año
df_resumen_anual = (
    secop_df.groupby(secop_df['fecha_de_firma'].dt.year)[
        ['valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago']
    ]
    .sum()
    .reset_index()
    .rename(columns={'fecha_de_firma': 'Año'})
)

# 3. Función ajustada a la escala larga (Colombia)
def formatear_cifras_colombia(valor):
    if pd.isna(valor):
        return valor
    
    val_abs = abs(valor)
    
    # Billones en Colombia (12 ceros -> 1.000.000.000.000)
    if val_abs >= 1e12:
        val = valor / 1e12
        sufijo = " B"
    # Millones en Colombia (6 ceros -> 1.000.000)
    elif val_abs >= 1e6:
        val = valor / 1e6
        sufijo = " M"
    else:
        val = valor
        sufijo = ""
    
    # Formateo visual (Punto = miles, Coma = decimales)
    if val.is_integer():
        texto_num = f"{int(val):,}"
    else:
        texto_num = f"{val:,.2f}"
    
    texto_es = texto_num.replace(',', 'X').replace('.', ',').replace('X', '.')
    
    return f"{texto_es}{sufijo}"

# 4. Aplicar formato
for col in ['valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago']:
    df_resumen_anual[col] = df_resumen_anual[col].apply(formatear_cifras_colombia)

display(df_resumen_anual)



,Año,valor_pagado,valor_facturado,valor_pendiente_de_pago
0,2019,"752.382,90 M","782.852,70 M","2,07 B"
1,2020,"2,32 B","2,45 B","4,29 B"
2,2021,"2,42 B","2,88 B","12.864,09 B"
3,2022,"3,41 B","4,60 B","4,60 B"
4,2023,"3,97 B","5,04 B","5,88 B"
5,2024,"5,40 B","6,36 B","7,22 B"
6,2025,"5,20 B","7,04 B","15,67 B"


In [448]:
"""Analisis de formato de las columnas financieras: valor_pagado, valor_facturado y valor_pendiente_de_pago. """

cols_financieras = ['valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago']

# Función para analizar qué caracteres especiales contiene cada columna
def auditar_formato_texto(df, columnas):
    diagnostico = []
    for col in columnas:
        serie_str = df[col].astype(str).str.strip()
        
        tiene_punto = serie_str.str.contains(r'\.', regex=True).sum()
        tiene_coma = serie_str.str.contains(r',', regex=True).sum()
        tiene_ambos = serie_str.str.contains(r'\.', regex=True) & serie_str.str.contains(r',', regex=True)
        solo_numeros = serie_str.str.match(r'^\d+$').sum()
        
        diagnostico.append({
            "Columna": col,
            "Tiene Punto (.)": tiene_punto,
            "Tiene Coma (,)": tiene_coma,
            "Tiene Ambos (. y ,)": tiene_ambos.sum(),
            "Solo Dígitos (Sin separadores)": solo_numeros,
            "Muestras de Texto": serie_str.unique()[:3].tolist() # Ver 3 ejemplos reales
        })
    return pd.DataFrame(diagnostico)

display(auditar_formato_texto(secop_df, cols_financieras))


,Columna,Tiene Punto (.),"Tiene Coma (,)","Tiene Ambos (. y ,)",Solo Dígitos (Sin separadores),Muestras de Texto
0,valor_pagado,0,0,0,196391,"[17896237, 389999685, 11970000]"
1,valor_facturado,0,0,0,196391,"[17896237, 389999685, 11970000]"
2,valor_pendiente_de_pago,0,0,0,196247,"[1679, 315, 0]"


In [449]:
"""Asimetria y Curtosis de las columnas financieras: valor_pagado, valor_facturado y valor_pendiente_de_pago. """
columnas_evaluadas = ['valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago']

# 1. Cálculo sobre la totalidad de los datos
resumen_estadistico = []

for col in columnas_evaluadas:
    skew_val = secop_df[col].skew()
    kurt_val = secop_df[col].kurt()  # Nota: Pandas calcula la curtosis de exceso (normal = 0)
    
    resumen_estadistico.append({
        "Columna": col,
        "Asimetría (Skewness)": round(skew_val, 4),
        "Curtosis (Kurtosis)": round(kurt_val, 4)
    })

df_métricas = pd.DataFrame(resumen_estadistico)

print("MÉTRICAS SOBRE DATOS COMPLETOS (INCLUYENDO ATÍPICOS)")
display(df_métricas)


MÉTRICAS SOBRE DATOS COMPLETOS (INCLUYENDO ATÍPICOS)


,Columna,Asimetría (Skewness),Curtosis (Kurtosis)
0,valor_pagado,168.3456,36616.2554
1,valor_facturado,137.6966,23803.6522
2,valor_pendiente_de_pago,443.1602,196390.9506


### Hallazgos

- fecha_de_inicio_del_contrato presenta 2440 registros nulos lo cual es el 1.24 % del total de este atributo. 
- Se encontraron 4 registros con documento_proveedor nulo o vacío
- en 2021 el valor pendiente de pago asciende a 13 mil billones de pesos sin embargo el Presupuesto General de la nación aprobado para ese año fue de $313,9 billones de pesos, lo que se traduze en 41 veces el presupuesto de todo un pais, el valor pagado mas el valor facturado es un total de 5.2 Billones.
- La precencia de los datos del 2021 hace que la Curtosis de las 3 variables numericas de como resultado una curtosis Leptocúrtica extrema.

## 4. Unicidad

In [450]:
""" 4.1 Análisis de unicidad de de id_contrato contra fecha_de_firma, fecha_de_inicio_del_contrato y fecha_de_fin_del_contrato """

total_registros = len(secop_df)
print(f"Total de registros: {total_registros:,}")

# TODO: ajustar a la(s) columna(s) que deberían identificar un registro de forma única
columnas_clave = [
    "id_contrato", "fecha_de_firma", "fecha_de_inicio_del_contrato", "fecha_de_fin_del_contrato"
]


duplicados = (
    secop_df
    .groupby(columnas_clave)
    .size()
    .reset_index(name="cantidad_duplicados")
    .query("cantidad_duplicados > 1")
    .sort_values("cantidad_duplicados", ascending=False)
)

num_duplicados = len(duplicados)
if num_duplicados > 0:
    print(f"Se encontraron {num_duplicados:,} llaves duplicadas:")
    display(duplicados.head(20))

    print("\nEjemplos de registros duplicados:")
    primera_llave = duplicados.iloc[0][columnas_clave]
    filtro = (secop_df[columnas_clave] == primera_llave.values).all(axis=1)
    display(secop_df[filtro])
else:
    columnas_str = " + ".join(columnas_clave)
    print(f"No se encontraron duplicados por la combinacion de  {columnas_str}")
    


Total de registros: 196,391
No se encontraron duplicados por la combinacion de  id_contrato + fecha_de_firma + fecha_de_inicio_del_contrato + fecha_de_fin_del_contrato


In [451]:
""" 4.2 Analisis de duplicados de la combinacion de fecha_de_inicio_del_contrato + id_contrato + proveedor_adjudicado. """

columnas_clave = ["fecha_de_inicio_del_contrato", "id_contrato", "proveedor_adjudicado"]

# Contar duplicados
duplicados = (
    secop_df
    .groupby(columnas_clave)
    .size()
    .reset_index(name="cantidad_duplicados")
    .query("cantidad_duplicados > 1")
    .sort_values("cantidad_duplicados", ascending=False)
)

# Resultado
num_duplicados = len(duplicados)
if num_duplicados > 0:
    print(f"Se encontraron {num_duplicados:,} combinaciones duplicadas de {columnas_clave}:")
    display(duplicados.head(20))

    print("\nEjemplo de registros duplicados:")
    primera_llave = duplicados.iloc[0][columnas_clave]
    filtro = (secop_df[columnas_clave] == primera_llave.values).all(axis=1)
    display(secop_df[filtro])
else:
    columnas_str = " + ".join(columnas_clave)
    print(f"✓ La combinación de {columnas_str} es única en todos los registros")


✓ La combinación de fecha_de_inicio_del_contrato + id_contrato + proveedor_adjudicado es única en todos los registros


### Decisiones
- 

## 5. Consistencia

In [452]:
# Análisis de consistencia de secop_df

# TODO: definir reglas de consistencia específicas del negocio y validarlas
# Ejemplo: fecha_inicio no debería ser posterior a fecha_fin
# inconsistentes = secop_df[secop_df["fecha_inicio"] > secop_df["fecha_fin"]]
# print(f"Registros inconsistentes: {len(inconsistentes):,}")
# display(inconsistentes)

# Ejemplo: llave foránea sin correspondencia en la tabla maestro
# sin_match = secop_df[~secop_df["id_referencia"].isin(OTRO_DATAFRAME["id_referencia"])]
# display(sin_match)


### Decisiones
- 